site/js/modules/webui/input.js

```diff
@@ -0,0 +1,146 @@
+import * as Autograder from '../autograder/base.js';
+
+// A general representation of a user input field.
+class Field {
+    constructor(
+            name, displayName, key,
+            {
+                type = 'string', required = false, placeholder = '',
+                attributes = '', labelBefore = true, extractInputFunc = undefined
+            } = {}) {
+        // The name of the field which will be used for targeting the field.
+        this.name = name;
+
+        // The display name that will be shown to the user.
+        this.displayName = displayName;
+
+        // The key used for the JSON field.
+        this.key = key;
+
+        // The HTML input type.
+        // For more information about possible types,
+        // see https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/input#input_types.
+        this.type = type;
+
+        // Flags the field requires user input.
+        this.required = required;
+
+        // The placeholder for the input.
+        this.placeholder = placeholder;
+
+        // Any additional attributes to the input field.
+        // If the field is required, the required attribute will be added automatically.
+        this.attributes = attributes;
+
+        // Determines the position of the HTML label with respect to the input.
+        this.labelBefore = labelBefore;
+
+        // A custom function for extracting the value from an input.
+        // By default, the input.value will be returned.
+        this.extractInputFunc = extractInputFunc;
+    }
+
+    validate() {
+        if ((this.name == undefined) || (this.name == '')) {
+            console.error(`Input field cannot have an empty name: ${JSON.stringify(this)}`);
+        }
+
+        if ((this.displayName == undefined) || (this.displayName == '')) {
+            console.error(`Input field cannot have an empty display name: ${JSON.stringify(this)}.`);
+        }
+
+        if ((this.key == undefined) || (this.key == '')) {
+            console.error(`Input field caKeynnot have an empty key: ${JSON.stringify(this)}.`);
+        }
+    }
+
+    toHTML() {
+        this.validate();
+
+        let attributes = this.attributes;
+        let displayName = this.displayName;
+        if (this.required) {
+            attributes += ' required';
+            displayName += ` <span class="required-color">*</span>`;
+        }
+
+        const label = `<label for="${this.name}">${displayName}</label>`;
+        const input = `<input type="${this.type}" id="${this.name}" name="${this.name}" placeholder="${this.placeholder}" ${attributes}/>`;
+
+        if (this.labelBefore) {
+            return `
+                <div class="input-field">
+                    ${label}
+                    ${input}
+                </div>
+            `;
+        } else {
+            return `
+                <div class="input-field">
+                    ${input}
+                    ${label}
+                </div>
+            `;
+        }
```

Comment:
I'm not a fan of the two different multiline strings.
I know they are not complex, but because of the lines, it looks like there are big chunks of different code when you are really just swapping some ordering.
What about putting them in an array and using reverse() if necessary?